### PR TITLE
[ISSUE #19410] remove request_options_provider from the …

### PR DESCRIPTION
**Pull request by maxi297**
…manifest

## What
Solves #19410 

## How
Remove it from the declarative_component_schema.yaml, re-generate the files, update the factory by still using InterpolatedRequestOptionProvider and update the sources.

## Recommended reading order
1. airbyte-cdk/python/airbyte_cdk/sources/declarative/declarative_component_schema.yaml: remove from manifest
2. airbyte-cdk/python/airbyte_cdk/sources/declarative/config_component_schema.json: remove from json schema
3. airbyte-cdk/python/airbyte_cdk/sources/declarative/models/declarative_component_schema.py: auto-generated files
4. airbyte-cdk/python/airbyte_cdk/sources/declarative/parsers/model_to_component_factory.py: cleanup factory
5. all the sources

## 🚨 User Impact 🚨
This is a breaking change as described by #19410 

TLDR:
* source-datascope: The request_options_provider was written but at the incorrect location rendering it useless
* source-monday: This source uses something deprecated and we'll need to find out how we can migrate/test it

A `X` indicates that I validated InterpolatedRequestOptionsProvider is created with at least one parameter that isn't `None`. A `~` indicates that there was something special to be looked at for this connector which prevented me from performing the migration. The assumptions is that I could not test completely the API when `NO SECRETS`. Some other errors were documented as well which could affect the completeness of the tests

X source-alpha-vantage/source_alpha_vantage/alpha_vantage.yaml
X source-ashby/source_ashby/ashby.yaml
&nbsp;&nbsp;NO SECRETS
X source-braze/source_braze/braze.yaml
X source-breezometer/source_breezometer/breezometer.yaml
X source-callrail/source_callrail/callrail.yaml
X source-chartmogul/source_chartmogul/chartmogul.yaml
X source-close-com/source_close_com/close_com.yaml
X source-coin-api/source_coin_api/coin_api.yaml
X source-coingecko-coins/source_coingecko_coins/coingecko_coins.yaml
X source-coinmarketcap/source_coinmarketcap/coinmarketcap.yaml
X source-convertkit/source_convertkit/convertkit.yaml
X source-datascope/source_datascope/datascope.yaml
&nbsp;&nbsp;request_options_provider is under retriever and not requester. It was removed for now
X source-emailoctopus/source_emailoctopus/emailoctopus.yaml
X source-facebook-pages/source_facebook_pages/facebook_pages.yaml
X source-gnews/source_gnews/gnews.yaml
X source-gocardless/source_gocardless/gocardless.yaml
&nbsp;&nbsp;NO SECRETS
X source-gong/source_gong/gong.yaml
X source-google-pagespeed-insights/source_google_pagespeed_insights/google_pagespeed_insights.yaml
&nbsp;&nbsp;NO SECRETS
X source-gutendex/source_gutendex/gutendex.yaml
X source-ip2whois/source_ip2whois/ip2whois.yaml
X source-mailersend/source_mailersend/mailersend.yaml
X source-mailjet-sms/source_mailjet_sms/mailjet_sms.yaml
&nbsp;&nbsp;NO SECRETS
X source-monday/source_monday/monday.yaml
&nbsp;&nbsp;CustomRequestOptionsProvider. A GraphQLRequestOptionsProvider was created
X source-n8n/source_n8n/n8n.yaml
X source-news-api/source_news_api/news_api.yaml
&nbsp;&nbsp;<Your plan permits you to request articles as far back as 2022-12-12, but you have requested 2022-10-26. You may need to upgrade to a paid plan.>
X source-newsdata/source_newsdata/newsdata.yaml
&nbsp;&nbsp;missing fields in secrets
X source-nytimes/source_nytimes/nytimes.yaml
X source-oura/source_oura/oura.yaml
&nbsp;&nbsp;NO SECRETS
X source-partnerstack/source_partnerstack/partnerstack.yaml
&nbsp;&nbsp;NO SECRETS
X source-pexels-api/source_pexels_api/pexels_api.yaml
X source-plausible/source_plausible/plausible.yaml
&nbsp;&nbsp;NO SECRETS
X source-pocket/source_pocket/pocket.yaml
X source-postmarkapp/source_postmarkapp/postmarkapp.yaml
X source-prestashop/source_prestashop/prestashop.yaml
X source-punk-api/source_punk_api/punk_api.yaml
X source-pypi/source_pypi/pypi.yaml
X source-recreation/source_recreation/recreation.yaml
X source-reply-io/source_reply_io/reply_io.yaml
&nbsp;&nbsp;<failed with status code 400 and error message Trial is expired or account is closed>
X source-rocket-chat/source_rocket_chat/rocket_chat.yaml
&nbsp;&nbsp;<Unable to connect to stream teams - Request to https://airbytesandbox.rocket.chat/api/v1/teams.list?count=1 failed with status code 400 and error message None>
X source-sendgrid/source_sendgrid/sendgrid.yaml
&nbsp;&nbsp;[OK based on Brian's comment, Alex ported it to low-code but it was deprioritized for release]This source is not a declarative one so I'm not sure what the sendgrid.yaml is used for...
X source-senseforce/source_senseforce/senseforce.yaml
X   source-sentry/source_sentry/sentry.yaml
&nbsp;&nbsp;[OK based on Brian's comment, Alex ported it to low-code but it was deprioritized for release]This source is not a declarative one so I'm not sure what the sendgrid.yaml is used for..
X source-smartengage/source_smartengage/smartengage.yaml
X source-sonar-cloud/source_sonar_cloud/sonar_cloud.yaml
X source-square/source_square/square.yaml
X source-the-guardian-api/source_the_guardian_api/the_guardian_api.yaml
X source-tmdb/source_tmdb/tmdb.yaml
X source-toggl/source_toggl/toggl.yaml
&nbsp;&nbsp;<Unable to connect to stream time_entries - Request to https://api.track.toggl.com/api/v9/me/time_entries?start_date=2022-10-01&end_date=2022-10-30 failed with status code 400 and error message start_date must not be earlier than 2022-10-13>
X source-tvmaze-schedule/source_tvmaze_schedule/tvmaze_schedule.yaml
X source-twitter/source_twitter/twitter.yaml
&nbsp;&nbsp;<Unable to connect to stream tweets - Request to https://api.twitter.com/2/tweets/search/recent?query=tesla&start_time=2022-12-16T10%3A00%3A00Z&end_time=2022-12-16T10%3A15%3A50Z&max_results=100 failed with status code 400 and error message Invalid 'start_time':'2022-12-16T10:00Z'. 'start_time' must be on or after 2023-01-06T16:04Z, Invalid 'end_time':'2022-12-16T10:15Z'. 'end_time' must be on or after 2023-01-06T16:04Z>
X source-vitally/source_vitally/vitally.yaml
&nbsp;&nbsp;NO SECRETS
X source-waiteraid/source_waiteraid/waiteraid.yaml
&nbsp;&nbsp;NO SECRETS
X source-wikipedia-pageviews/source_wikipedia_pageviews/wikipedia_pageviews.yaml
X source-woocommerce/source_woocommerce/woocommerce.yaml
X source-workable/source_workable/workable.yaml
&nbsp;&nbsp;NO SECRETS
X source-zapier-supported-storage/source_zapier_supported_storage/zapier_supported_storage.yaml
X source-zenloop/source_zenloop/zenloop.yaml

### CODE REVIEW

File: airbyte-cdk/python/airbyte_cdk/sources/declarative/declarative_component_schema.yaml [hunk 1]
```diff
@@ -654,10 +637,30 @@ definitions:
               - GET
               - POST
         default: GET
-      request_options_provider:
+      request_body_data:
```
Comment (brianjlai) [inline airbyte-cdk/python/airbyte_cdk/sources/declarative/declarative_component_schema.yaml:640]: Can you add `description` to each of these fields? We're trying to get in the habit of having every field marked this way, so anytime we edit existing fields we should add them if it doesn't already exist. The existing component might have the parameters described in the docstring

File: airbyte-cdk/python/airbyte_cdk/sources/declarative/declarative_component_schema.yaml [hunk 2]
```diff
@@ -654,10 +637,30 @@ definitions:
...
         anyOf:
-          - "$ref": "#/definitions/CustomRequestOptionsProvider"
-          - "$ref": "#/definitions/InterpolatedRequestOptionsProvider"
+          - type: string
+          - type: object
+            additionalProperties:
+              type: string
+      request_body_json:
+        anyOf:
+          - type: string
```
Comment (brianjlai) [inline airbyte-cdk/python/airbyte_cdk/sources/declarative/declarative_component_schema.yaml:648]: Since these are a complex object that we reuse a little bit, might be worth defining a new property called RequestInput. I think we did this for PrimaryKey as some prior art

**Review by brianjlai — COMMENTED** [review]
just a few initial comments on the draft you have out. Looks great so far! I think you mentioned the main thing to add was addressing source-monday which used a custom interpolated request options component. I'll re-review afterwards, but so far don't see any issues w/ the code flow so far.

Also, if you used a script to migrate existing connectors, can you include that in the
`/python/bin/` folder. Just useful to document it and we can delete them after we move to beta.

File: airbyte-cdk/python/unit_tests/sources/declarative/test_manifest_declarative_source.py [hunk 3]
```diff
@@ -182,7 +182,7 @@ def test_manifest_with_spec(self, construct_using_pydantic_models):
                         "requester": {
                             "path": "/v3/marketing/lists",
                             "authenticator": {"type": "BearerAuthenticator", "api_token": "{{ config.apikey }}"},
-                            "request_parameters": {"page_size": 10},
+                            "request_parameters": {"page_size": "{{ 10 }}"},
```
[the remark this answers is not in the document]
Comment (maxi297) [reply airbyte-cdk/python/unit_tests/sources/declarative/test_manifest_declarative_source.py:185]: I confirm that this is working as well

File: airbyte-cdk/python/airbyte_cdk/sources/declarative/declarative_component_schema.yaml (hunk 1 above)
Comment (maxi297) [reply airbyte-cdk/python/airbyte_cdk/sources/declarative/declarative_component_schema.yaml:640]: Done

File: airbyte-cdk/python/airbyte_cdk/sources/declarative/declarative_component_schema.yaml (hunk 2 above)
Comment (maxi297) [reply airbyte-cdk/python/airbyte_cdk/sources/declarative/declarative_component_schema.yaml:648]: If I understand your point properly, I'm a little bit reluctant to do that. My reasoning is that I feel like those fields are not representing the same thing and therefore might not evolve in the same way. One example I see of this is https://github.com/airbytehq/airbyte/issues/20734 where `request_body_json` would have a very different type definition than the others. I think we can argue the same for the other fields. Therefore, I feel like re-using a complex type for the different `request_*` field might lead to more work later when these will evolve differently. 

Let me know if this makes sense for you. If not, just say it and I'll define the `RequestInput` as suggested

**Comment by maxi297** [comment]
> Also, if you used a script to migrate existing connectors, can you include that in the `/python/bin/` folder. Just useful to document it and we can delete them after we move to beta.

I didn't create a script to do this because there were a couple of edge-cases involving refs and testing that I thought would have been more work to figure out

File: airbyte-cdk/python/airbyte_cdk/sources/declarative/declarative_component_schema.yaml (hunk 2 above)
Comment (brianjlai) [reply airbyte-cdk/python/airbyte_cdk/sources/declarative/declarative_component_schema.yaml:650]: Fair enough, I agree with your rationale and we can leave as is! Thanks for adding the descriptions too!

File: airbyte-integrations/connectors/source-monday/source_monday/graphql_requester.py [hunk 4]
```diff
@@ -109,3 +109,9 @@ def get_request_params(
             page=next_page_token and next_page_token[self.NEXT_PAGE_TOKEN_FIELD_NAME],
         )
         return {"query": f"query{{{query}}}"}
+
+    # We are using an LRU cache in should_retry() method which requires all incoming arguments (including self) to be hashable.
+    # Dataclasses by default are not hashable, so we need to define __hash__(). Alternatively, we can set @dataclass(frozen=True),
+    # but this has a cascading effect where all dataclass fields must also be set to frozen.
+    def __hash__(self):
```
Comment (brianjlai) [inline airbyte-integrations/connectors/source-monday/source_monday/graphql_requester.py:116]: nice find to address this

**Review by brianjlai — APPROVED** [review]
Looks good to me! Since there's no migration, we can probably just go ahead with including these changes when we merge things in. 

And can you re run if you haven't already `./gradlew validateSourceYamlManifest` to validate and ensure that `read` works for `source-monday` since that one had an interesting edge case to account for?

**Comment by maxi297** [comment]
> And can you re run if you haven't already ./gradlew validateSourceYamlManifest to validate and ensure that read works for source-monday since that one had an interesting edge case to account for?

Both `./gradlew validateSourceYamlManifest` and `./gradlew --no-daemon --scan :airbyte-integrations:connectors:source-monday:integrationTest` work!
